share/core.go

```diff
@@ -0,0 +1,315 @@
+type PriPoly struct {
+	g      abstract.Group    // Cryptographic group
+	coeffs []abstract.Scalar // Coefficients of the polynomial
+}
+
+// NewPriPoly creates a new secret sharing polynomial for the cryptographic
+// group g, the secret sharing threshold t, and the secret to be shared s.
+func NewPriPoly(g abstract.Group, t int, s abstract.Scalar, rand cipher.Stream) *PriPoly {
+	coeffs := make([]abstract.Scalar, t)
+	coeffs[0] = s
+	if coeffs[0] == nil {
+		coeffs[0] = g.Scalar().Pick(rand)
+	}
+	for i := 1; i < t; i++ {
+		coeffs[i] = g.Scalar().Pick(rand)
+	}
+	return &PriPoly{g, coeffs}
+}
+
+// Threshold returns the secret sharing threshold.
+func (p *PriPoly) Threshold() int {
+	return len(p.coeffs)
+}
+
+// GetSecret returns the shared secret p(0), i.e., the constant term of the polynomial.
+func (p *PriPoly) GetSecret() abstract.Scalar {
+	return p.coeffs[0]
+}
+
+// Eval computes the private share v = p(i).
+func (p *PriPoly) Eval(i int) *PriShare {
+	xi := p.g.Scalar().SetInt64(1 + int64(i)) // x-coordinate of this share
+	v := p.g.Scalar().Zero()
+	for j := p.Threshold() - 1; j >= 0; j-- {
+		v.Mul(v, xi)
+		v.Add(v, p.coeffs[j])
+	}
+	return &PriShare{i, v}
+}
+
+// Shares creates a list of n private shares p(1),...,p(n).
+func (p *PriPoly) Shares(n int) []*PriShare {
+	shares := make([]*PriShare, n)
+	for i := 0; i < n; i++ {
+		shares[i] = p.Eval(i)
+	}
+	return shares
+}
+
+// Add computes the component-wise sum of the polynomials p and q and returns it
+// as a new polynomial.
+func (p *PriPoly) Add(q *PriPoly) (*PriPoly, error) {
+
+	if p.g != q.g {
+		return nil, errorGroups
+	}
+
+	if p.Threshold() != q.Threshold() {
+		return nil, errorCoeffs
+	}
+
+	t := p.Threshold()
+	coeffs := make([]abstract.Scalar, t)
+	for i := 0; i < t; i++ {
+		coeffs[i] = p.g.Scalar().Add(p.coeffs[i], q.coeffs[i])
+	}
+
+	return &PriPoly{p.g, coeffs}, nil
+}
+
+// Equal checks equality of two secret sharing polynomials p and q.
+func (p *PriPoly) Equal(q *PriPoly) bool {
+
+	if p.g.String() != q.g.String() {
+		return false
+	}
+
+	for i := 0; i < p.Threshold(); i++ {
+		if !p.coeffs[i].Equal(q.coeffs[i]) {
+			return false
+		}
+	}
+
+	return true
+}
+
+// Commit creates a public commitment polynomial for the given base point b or
+// the standard base if b == nil.
+func (p *PriPoly) Commit(b abstract.Point) *PubPoly {
+	t := p.Threshold()
+	commits := make([]abstract.Point, t)
+	for i := 0; i < t; i++ {
+		commits[i] = p.g.Point().Mul(b, p.coeffs[i])
+	}
+	return &PubPoly{p.g, b, commits}
+}
+
+// RecoverSecret reconstructs the shared secret p(0) using Lagrange interpolation.
+func RecoverSecret(g abstract.Group, shares []*PriShare, t int, n int) (abstract.Scalar, error) {
+
+	isBad := func(s *PriShare) bool {
+		return s == nil || s.V == nil || s.I < 0 || n <= s.I
+	}
+
+	c := 0
+	x := make(map[int]abstract.Scalar)
+	for _, s := range shares {
+		if isBad(s) {
+			continue
+		}
+		x[s.I] = g.Scalar().SetInt64(1 + int64(s.I))
```

Comment:
You should  not need to declare the `isBad` function really, since you only need here. 
Using like that:
```
for i, s := range shares {
     if s == nil || s.V == nil || s.I < 0 || n <= s.I {
         continue
     }
     x[i] = g.Scalar().SetInt64(1 + int64(s.I))
}
```
then, later, you can directly iterate over *only* the good shares (vs *all* the shares):
```
for i,si := range x { // line 150
    // no need for isBad here
    for j,sj := range x {
        if j == i { // no need for isBad here
            continue
        }
        ... // same thing here
    }
}
```

Sorry maybe I confused you with my suggestion of using a separate function. In fact, my idea was to use a common function that retrieves the good shares whether using `PriShare`s or `PubShare`s but when I tried to implement it did not look as good as I thought :D